### PR TITLE
[SPARK-40697][SQL] Add read-side char padding to cover external data files

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

The current char/varchar feature relies on the data source to take care of all the write paths and ensure the char/varchar semantic (length check, string padding). This is good for the read performance, but has problems if some write paths did not follow the char/varchar semantic. e.g. a parquet table can be written by old Spark versions that do not have char/varchar type, or by other systems that do not recognize Spark char/varchar type.

This PR adds read-side string padding for the char type, so that we can still guarantee the char type semantic if the underlying data is valid (not over length). Char type is rarely used for legacy reasons and the perf doesn't matter that much, correctness is more important here. People can still disable read-side padding via a config if they are sure the data was written properly, such as benchmarks.

Note, we don't add read-side length check as varchar type is widely used and we don't want to introduce perf regression for the common case. Another reason is it's better to avoid invalid data at the write side, and read-side check won't help much.

### Why are the changes needed?

to better enforce char type semantic

### Does this PR introduce _any_ user-facing change?

Yes. Now Spark can still return padding char type values correctly even if the data source writer wrote the char type value without padding.

### How was this patch tested?

updated tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yaooqinn @gengliangwang

**Comment by yaooqinn** [comment]
some of the string functions may be affected, such as `length`, `rpad`, `concat` etc

**Comment by cloud-fan** [comment]
@yaooqinn any thoughts on the problems? I thought the only problem is perf, but char type is rarely used anyway.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 1]
```diff
@@ -52,4 +52,15 @@ public static UTF8String varcharTypeWriteSideCheck(UTF8String inputStr, int limi
...
   }
+
+  public static UTF8String readSidePadding(UTF8String inputStr, int limit) {
+    int numChars = inputStr.numChars();
+    if (numChars == limit) {
+      return inputStr;
+    } else if (numChars < limit) {
+      return inputStr.rpad(limit, SPACE);
+    } else {
+      return inputStr;
```
Comment (gengliangwang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:63]: Shall we throw an exception if the `numChars` exceeds `limit`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+/**
+ * This rule performs string padding for char type.
+ *
+ * When reading values from column/field of type CHAR(N), right-pad the values to length N, if the
+ * read-side padding config is turned on.
+ *
+ * When comparing char type column/field with string literal or char type column/field,
+ * right-pad the shorter one to the longer length.
+ */
+object ApplyCharTypePadding extends Rule[LogicalPlan] {
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala:39]: Why do we move the rule under the datasources package?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:63]: We need a test case for this branch, no matter what the expected behavior is.

**Comment by yaooqinn** [comment]
would the result of  concat('abc', 'defg') be changed from 'abcdefg' to 'abc defg'?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ApplyCharTypePadding.scala:39]: It was there at the first place: https://github.com/apache/spark/commit/5cfbdddefe0753c3aff03f326b31c0ba8882b3a9#diff-da01d4c9147810ef330a7e70ad197fda5e768ad17558176e686d3e63139172b5

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:63]: We have tests for it, and I didn't add length check to match VARCHAR. Note: we add can read-side length check for both CHAR and VARCHAR, but that's bad for perf is VARCHAR is common.

**Comment by cloud-fan** [comment]
@yaooqinn  It will not, as read-side padding only applies to char type.

**Comment by yaooqinn** [comment]
can we add some unit tests for `char + string funcs`?

**Comment by cloud-fan** [comment]
@yaooqinn you are probably confused by 2 kinds of char paddings:
1. column value padding. This is done at the write side, and this PR adds read-side padding as well. It doesn't matter how/where the column is used. The semantic is the column value was padded while writing.
2. char comparison padding. This is done at the read side, and this PR doesn't touch it. When comparsing a char column with a string literal, we pad the shorter one to the longer length.

**Comment by yaooqinn** [comment]
```scala
  withTempPath { dir =>
      withTable("t") {
        sql("SELECT '12' as col1, '12' as col2").write.format(format).save(dir.toString)
        sql(s"CREATE TABLE t (col1 char(3), col2 varchar(3)) using $format")
        sql(s"ALTER TABLE t SET LOCATION '$dir'")
        checkAnswer(spark.sql("select concat(col1, col2) from t"), Row("12 12"))
      }
    }
```

I mean a test case like the above. W/ this PR, the 1st case below is a behavior change that needs a doc or unit test.

```
// length fits
12 + 12 = 12<space>12
// length overflow w/ significant characters
1234 + 12 = 1234
// length overflow w/ spaces
12<space><space> + 12 = 12<space><space>34
```

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 3]
```diff
@@ -830,14 +828,12 @@ class FileSourceCharVarcharTestSuite extends CharVarcharTestSuite with SharedSpa
...
-          sql(s"CREATE TABLE t (col $typ(2)) using $format")
-          sql(s"ALTER TABLE t SET LOCATION '$dir'")
-          checkAnswer(spark.table("t"), Row("12"))
-        }
+    withTempPath { dir =>
+      withTable("t") {
+        sql("SELECT '12' as col1, '12' as col2").write.format(format).save(dir.toString)
+        sql(s"CREATE TABLE t (col1 char(3), col2 varchar(3)) using $format")
+        sql(s"ALTER TABLE t SET LOCATION '$dir'")
+        checkAnswer(spark.table("t"), Row("12 ", "12"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:836]: @yaooqinn I think this test covers the behavior change you mentioned? And I'd treat it as a bug fix, as the previous result didn't follow CHAR type semantic.

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:836]: make sense to me

**Review by gengliangwang — APPROVED** [review]
+1

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for review, merging to master!

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @cloud-fan , @gengliangwang , @yaooqinn .

**Pull request closed without merging**